trieste/space.py

```diff
@@ -920,17 +917,242 @@ def __init__(self, spaces: Sequence[SearchSpace], tags: Optional[Sequence[str]]
+        return [self.get_subspace(tag).lower for tag in self.subspace_tags]
+
+    @property
+    def subspace_upper(self) -> Sequence[TensorType]:
+        """The highest values taken by each space dimension, in the same order as specified when
+        initializing the space."""
+        return [self.get_subspace(tag).upper for tag in self.subspace_tags]
+
+    @property
+    def subspace_tags(self) -> tuple[str, ...]:
+        """Return the names of the subspaces contained in this space."""
+        return self._tags
+
+    @property
+    def subspace_dimension(self) -> Sequence[TensorType]:
+        """The number of inputs in each subspace, in the same order as specified when initializing
+        the space."""
+        return [self.get_subspace(tag).dimension for tag in self.subspace_tags]
+
+    def get_subspace(self, tag: str) -> SearchSpace:
+        """
+        Return the domain of a particular subspace.
+
+        :param tag: The tag specifying the target subspace.
+        :return: Target subspace.
+        """
+        tf.debugging.assert_equal(
+            tag in self.subspace_tags,
+            True,
+            message=f"""
+                Attempted to access a subspace that does not exist. This space only contains
+                subspaces with the tags {self.subspace_tags} but received {tag}.
+                """,
+        )
+        return self._spaces[tag]
+
+    def subspace_sample(self, num_samples: int, seed: Optional[int] = None) -> Sequence[TensorType]:
+        """
+        Sample randomly from the space by sampling from each subspace
+        and returning the resulting samples in the same order as specified when initializing
+        the space.
+
+        :param num_samples: The number of points to sample from each subspace.
+        :param seed: Optional tf.random seed.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly,
+            from each search subspace with shape '[num_samples, D]' , where D is the search space
+            dimension.
+        """
+        tf.debugging.assert_non_negative(num_samples)
+        if seed is not None:  # ensure reproducibility
+            tf.random.set_seed(seed)
+        return [self._spaces[tag].sample(num_samples, seed=seed) for tag in self._tags]
+
+    def __eq__(self, other: object) -> bool:
+        """
+        :param other: A search space.
+        :return: Whether the search space is identical to this one.
+        """
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self._tags == other._tags and self._spaces == other._spaces
+
+
+class TaggedMultiSearchSpace(CollectionSearchSpace):
+    r"""
+    A :class:`SearchSpace` consisting of a collection of multiple :class:`SearchSpace` objects,
+    each with a unique tag. This class provides functionality for accessing each individual space.
+
+    Note that the individual spaces are not combined in any way, however all subspaces should have
```

Comment:
Add longer explanation of how TaggedMultiSearchSpace can be used? At the moment it's a bit cryptic!

Also maybe have TaggedProductSearchSpace come before this in the file as it's conceptually simpler and used more widely?

Reply:
I have swapped the order of the classes and updated the docstring for `TaggedMultiSearchSpace`. Is that clearer? I can add a code example if that helps.